[NEEDS REVIEW] Throw/Pick Up Drones/Various Animals! (Pick up /living/ framework + Throw mob_holders)

Ziiro opened the pull request:
:cl: ChuckTheSheep, FlattestGuitar, Ziiro
add: You can now pick up (and wear!) certain simple mobs/pais/monkeys with alt+click
/:cl:

This PR is mostly made up of https://github.com/tgstation/tgstation/pull/35532 but it also fixes worn mob breathing/atmos (thanks FlattestGuitar) and makes some adjustments for futureproofing so that carbons could be held if that's a wanted feature.

![2018-09-11_11-13-58](https://user-images.githubusercontent.com/26494679/45375989-ddd07600-b5b3-11e8-8ad2-6317fa08d705.gif)

Ziiro committed:
mob_holder framework

MrDoomBringer commented:
does this reactor climbing on people's backs?

Ziiro commented:
No

Ziiro committed:
whoops disposals work again honk honk

outbools commented on code/modules/mob/living/inhand_holder.dm:
```diff
@@ -5,80 +5,140 @@
+		name = M.name
+		desc = M.desc
+
+	if(_can_head_override)
```
🤔

Wilchenx commented:
@Ziiro https://github.com/tgstation/tgstation/pull/25114

Ziiro commented:
Man going around that would be a fucking mess. I wonder if they should be a neck item instead.

Ziiro committed:
why was that even renamed what the christ

ExcessiveUseOfCobblestone commented:
@Ziiro it's `dynamic_hair_suffix` for the first problem

Ziiro committed:
thanks cobby

Ziiro commented:
This is ready for judgement

Ziiro committed:
Merge branch 'master' into mob-holder-1

# Conflicts:
#	code/modules/mob/living/carbon/monkey/monkey.dm

Tlaltecuhtli commented:
assault pocket monkey time

Ziiro commented:
they're a bulky item, so the only way you can store them is in a BOH. Also they can't breathe in a BOH. That being said, when you throw a mob holder it counts as *throwing a mob*. So you could just have a backpack of holding full of monkeys you throw at people to knock them down.

davethwave commented:
What about other players? Can you just stuff people or corpses in Boh's?

Ziiro commented:
No. In this version* of this PR, there is no way to pick up anything other than Monkeys, Station Pets, Drones, and pAIs.

*In the Oracle version, I made getting dwarfed in genetics pick up-able. It's total hilarious/fun, but I feel like there's some balance repercussions that would arise out of that on a more robust playerbase that doesn't occur on Oracle.

Fox-McCloud commented on code/game/machinery/suit_storage_unit.dm:
```diff
@@ -305,6 +305,9 @@
 /obj/machinery/suit_storage_unit/attackby(obj/item/I, mob/user, params)
 	if(state_open && is_operational())
+		if(istype(I, /obj/item/clothing/head/mob_holder))
+			to_chat(user, "<span class='warning'>That doesn't fit in the unit like that! Try another method!</span>")
```
This is gross, and snowflakey. This should be expanded to be something that is either a flag or a var on an object that, if set, prevents it from being placed inside this.

Fox-McCloud commented on code/game/machinery/washing_machine.dm:
```diff
@@ -213,6 +213,10 @@
 		update_icon()
 		return
 
+	if(istype(W, /obj/item/clothing/head/mob_holder))
```
same here.

Fox-McCloud commented on code/modules/food_and_drinks/kitchen_machinery/deep_fryer.dm:
```diff
@@ -90,6 +90,9 @@ God bless America.
 	if(istype(I, /obj/item/reagent_containers/food/snacks/deepfryholder))
 		to_chat(user, "<span class='userdanger'>Your cooking skills are not up to the legendary Doublefry technique.</span>")
 		return
+	if(istype(I,/obj/item/clothing/head/mob_holder))
```
and here.

Fox-McCloud commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -150,19 +150,26 @@
-					add_logs(src, throwable_mob, "thrown", addition="grab from tile in [AREACOORD(start_T)] towards tile at [AREACOORD(end_T)]")
+	var/mob/living/throwable_mob
+
+	if(istype(I, /obj/item/clothing/head/mob_holder))
```
This is also pretty gross; snowflaking behavior out is a blatant violation of OOP.

Fox-McCloud commented on code/modules/mob/living/inhand_holder.dm:
```diff
@@ -5,80 +5,141 @@
+	return holder
+
+/mob/living/proc/mob_try_pickup(mob/living/user)
+	if(!ishuman(user) || !src.Adjacent(user) || user.incapacitated() || !can_be_held)
```
no implied `src`. Here or anywhere else unless you ever change `src` or have poorly named variables.

Fox-McCloud commented on code/modules/recycling/disposal/bin.dm:
```diff
@@ -105,8 +105,17 @@
 /obj/machinery/disposal/proc/place_item_in_disposal(obj/item/I, mob/user)
-	I.forceMove(src)
-	user.visible_message("[user.name] places \the [I] into \the [src].", "<span class='notice'>You place \the [I] into \the [src].</span>")
+	if(istype(I, /obj/item/clothing/head/mob_holder))
```
❄️

Fox-McCloud commented on code/modules/research/destructive_analyzer.dm:
```diff
@@ -152,3 +152,8 @@ Note: Must be placed within 3 tiles of the R&D Console
+
+/obj/machinery/rnd/destructive_analyzer/attackby(obj/item/I, mob/user)
+	if(istype(I, /obj/item/clothing/head/mob_holder))
+		to_chat(user, "<span class='warning'>The [src.name] smartly refuses \the [I].</span>")
```
no `\the`; BYOND handles this already. also no implied `src` either.

Fox-McCloud commented:
I kinda wonder if mob holders should be a clothing subtype at all. As far as I know, something doens't necessarily have to be a clothing/head subtype to be worn on the head and for it to be rendered properly.

As far as I know, this can all be controlled with clothing flags.

The fact that so many clothing exceptions have to be made inclines me to believe that them being a clothing subtype may not be the best thing.

Kagus commented:
Paper can be worn on the head, and the chaplain's atheist fedora can also be worn on the head, despite not actually being a hat.

Doppeltongue commented:
If manlets don't want to get tossed into the SM, they should get some mutadone.

Ziiro commented:
@Fox-McCloud I could switch it over to a basic obj type? I think the warning cone is probably a good example of what I have in mind.

also with regards to ❄️: I could make it an item trait, but I'm not sure what all of those things have in common I could put under one sort of flag (`NO_DESTROY`?). I could make it a set of flags with each device having a check: i.e, `NO_FRY`, `NO_WASH`, `NO_ANALYZE`, `NO_SUIT_STORAGE`, but I'm not actually sure that's any better. Is it?

ExcessiveUseOfCobblestone commented:
Don't think it should be a component, a living flag

Ziiro commented:
It's already a living var. kevinz000 put that in a different animal pickup PR

https://github.com/tgstation/tgstation/blob/b9c5bad6055496cc38dba541902b55ae1c50234d/code/modules/mob/living/living_defines.dm#L106

I'll fix the ❄️ I just need to know what actually works. A singular item flag, or multiple?

I figure I can re-organize it under `obj/mob_holder` and still make it able to be worn (Thanks for the Atheist Fedora blueprint on that)

Ziiro committed:
moves to abstract and removes clothing/head

Ziiro commented:
why the fuck didn't anyone just tell me to use abstract, christ jesus it's that easy (for most of the problems)

ExcessiveUseOfCobblestone commented:
Not everyone can be an epic silicon like your avatar, anon

Ziiro committed:
Conflict resolution, maybe?

# Conflicts:
#	code/modules/mob/living/carbon/carbon.dm
#	code/modules/mob/living/simple_animal/friendly/sloth.dm

Ziiro committed:
ebin

Ziiro committed:
giving back those spaces

Ziiro committed:
okay how about a few more spaces

TheDreamweaver commented:
Can we get a review up in here.

GrayRachnid commented:
alt+click for this seems very hacky.

Ziiro commented on code/modules/mob/living/inhand_holder.dm:
```diff
@@ -1,84 +1,145 @@
+		return FALSE
+	visible_message("<span class='warning'>[user] starts picking up [src].</span>", \
+					"<span class='userdanger'>[user] starts picking you up!</span>")
+	if(!do_after(user, 20, target = src))
```
Shit I forgot I can use `2 SECONDS` here, neat.

Ziiro committed:
2 SECONDS

Ziiro committed:
Merge branch 'master' into mob-holder-1

# Conflicts:
#	code/modules/mob/living/simple_animal/friendly/dog.dm

nicbn commented:
alt+click sucks
As ctrl+click pull already does the job for grab intent, make grab intent pick the mob up (if it can't be picked up, it behaves like pull)

Ziiro commented:
Problem is I'd like to have this framework available for carbons as well. If I tie it to grab intent, that will definitely suck.

ExcessiveUseOfCobblestone commented:
whats the issue with altclick again?

Kagus commented:
> whats the issue with altclick again?

We don't approve of alt culture around here.

nicbn commented:
It's not intuitive and it is already used for a lot of unrelated things

ExcessiveUseOfCobblestone commented:
picking up is the _alternative_ to furthering the grab into choke or whatever so I don't find that not intuitive. This is probably one of the few things that DOES make sense to toss on alt.

Having normal grab function conflict with this for the sole sake of appeasing non-alt users is ???. At the very least, it should be a singular control so we aren't further taking up click possibilities.

It's nice to know this PR is so good the only thing left to bikeshed is people's preferences on what to click though.

Kagus commented:
Hey, nobody can resist a good game of dwarf tossing. Not even the dwarfs!

Especially not the dwarfs.

Ziiro commented:
Going on hiatus from SS13 again, if someone else wants to bother with this they have my support.
